glom/grouping.py

```diff
@@ -0,0 +1,197 @@
+    """
+    recurse = lambda spec: scope[glom](target, spec, scope)
+    acc2 = scope[ACC2]  # current acuumulator support structure
+    if callable(getattr(spec, "agg", None)):
+        return spec.agg(target, acc2)
+    if type(spec) is dict:
+        acc = scope[ACC]  # current accumulator
+        for keyspec, valspec in spec.items():
+            key = recurse(keyspec)
+            if key not in acc:
+                acc[key] = _mk_acc(valspec)
+                acc2[key] = {}
+            scope[ACC] = acc[key]
+            scope[ACC2] = acc2[key]
+            acc[key] = recurse(valspec)
+        return acc
+    elif type(spec) is list:
+        acc = scope[ACC]  # current accumulator
+        for valspec in spec:
+            assert type(valspec) is not dict
+            # dict inside list is not valid
+            acc.append(recurse(valspec))
+        return acc
+    elif callable(spec):
+        return spec(target)
+    raise TypeError("not a valid spec")
+
+
+def _mk_acc(spec):
+    """
+    make an acculumator for a given spec
+    """
+    if type(spec) is dict:
+        return {}
+    if type(spec) is list:
+        return []
+    return None
+
+
+class First(object):
+    """
+    holds onto the first value
+
+    >>> glom([1, 2, 3], Group(First()))
+    1
+    """
+    # TODO: this could be optimized with a "done" flag
+    # set in acc2; dicts could check if all children are
+    # "done" set their own done flag, roll it up all the way
+    # to the root
+    __slots__ = ()
+
+    def agg(self, target, acc2):
+        if self not in acc2:
+            acc2[self] = target
+        return acc2[self]
+
+
+class Avg(object):
+    """
+    takes the numerical average of all values;
+    raises exception on non-numeric value
+
+    >>> glom([1, 2, 3], Group(Avg()))
+    2.0
+    """
+    __slots__ = ()
+
+    def agg(self, target, acc2):
+        if self not in acc2:
+            acc2[self] = [0, 0.0]
+        acc2[self][0] += target
+        acc2[self][1] += 1
+        return acc2[self][0] / acc2[self][1]
+
+
+class Sum(object):
+    """
+    takes the sum of all values;
+    raises exception on values incompatible with addition operator
+
+    >>> glom([1, 2, 3], Group(Sum()))
+    6
+    """
+    __slots__ = ()
+
+    def agg(self, target, acc2):
+        if self not in acc2:
+            acc2[self] = 0
+        acc2[self] += target
+        return acc2[self]
+
+
+class Max(object):
+    """
+    takes the maximum of all values;
+    raises exception on values that are not comparable
+
+    >>> glom([1, 2, 3], Group(Max()))
+    3
+    """
+    __slots__ = ()
+
+    def agg(self, target, acc2):
+        if self not in acc2:
+            acc2[self] = target
+        if target > acc2[self]:
+            acc2[self] = target
+        return acc2[self]
+
+
+class Min(object):
+    """
+    takes the minimum of all values;
+    raises exception on values that are not comparable
+
+    >>> glom([1, 2, 3], Group(Min()))
```

Comment:
Should Min() just auto-switch into the mode? Or do you want to auto-guard against that evaluation for multi-pass glomming?

Reply:
this is a very deep conversation....

right now an aggregate and a spec are different objects; if we make them the same, then the recursive structure inside the group mode changes a bit

Reply:
.......... okay...... mayyyyyyybeeeeeeee there's a way

Reply:
but one of the things I was trying to guard against was accidentally crossing them up